delivery/services/organise_service.py

```diff
@@ -0,0 +1,186 @@
+
+import logging
+import os
+import time
+
+from delivery.exceptions import ProjectAlreadyOrganisedException
+
+from delivery.models.project import RunfolderProject
+from delivery.models.runfolder import Runfolder
+from delivery.models.sample import Sample, SampleFile
+from delivery.services.file_system_service import FileSystemService
+
+log = logging.getLogger(__name__)
+
+
+class OrganiseService(object):
+    """
+    Starting in this context means organising a runfolder in preparation for a delivery. Each project on the runfolder
+    will be organised into its own separate directory. Sequence and report files will be symlinked from their original
+    location.
+    This service handles that in a synchronous way.
+    """
+
+    def __init__(self, runfolder_service, file_system_service=FileSystemService()):
+        """
+        Instantiate a new OrganiseService
+        :param runfolder_service: an instance of a RunfolderService
+        :param file_system_service: an instance of FileSystemService
+        """
+        self.runfolder_service = runfolder_service
+        self.file_system_service = file_system_service
+
+    def organise_runfolder(self, runfolder_id, lanes, projects, force):
+        """
+        Organise a runfolder in preparation for delivery. This will create separate subdirectories for each of the
+        projects and symlink all files belonging to the project to be delivered under this directory.
+
+        :param runfolder_id: the name of the runfolder to be organised
+        :param lanes: if not None, only samples on any of the specified lanes will be organised
+        :param projects: if not None, only projects in this list will be organised
+        :param force: if True, a previously organised project will be renamed with a unique suffix
+        :raises ProjectAlreadyOrganisedException: if project has already been organised and force is False
+        :return: a Runfolder instance representing the runfolder after organisation
+        """
+        runfolder = self.runfolder_service.find_runfolder(runfolder_id)
+        projects_on_runfolder = self.runfolder_service.find_projects_on_runfolder(
+            runfolder,
+            only_these_projects=projects)
+        organised_projects = []
+        for project in projects_on_runfolder:
```

Comment:
A question on this. If I read this code correctly it will raise if it fails to organize any of the projects. Could this leave behind a stale state, where N project have been successfully organized, and then there is a failure for project N+1, and this will crash the whole thing without rolling back?

Perhaps it would be cleaner to separate the two things into phases. Where you 1) gather info on if everything is alright to organize, 2) organize it all in one go? Since rolling back will probably be problematic (since this involves file system operations) doing thing this way would perhaps at least mitigate the risk of this happening.

Reply:
I saw that you wrote a comment about that in the description now that I re-read it.

Reply:
yes, well spotted! I noted it as a todo but I think I'll fix it right away and update this PR.